File: bindings/python/src/kriya/jsonschema.py
"""Standards-compliant JSON Schema exporter (draft 2020-12 clean).

kriya's internal :class:`ParameterSchema` uses a compact custom shape where ``required`` is a
per-property boolean and ``items`` may be a bare type string. MCP clients and strict validators
(e.g. the Anthropic tool API) expect standard JSON Schema where ``required`` is an array of names
on the *parent* object and ``items`` is always a schema object. This converts between the two,
byte-for-byte matching kriya-core's ``jsonschema.ts`` so a Python app emits identical tool schemas.
"""

from __future__ import annotations

from typing import Any, Dict, List

from .types import ParameterSchema


def to_json_schema(schema: ParameterSchema) -> Dict[str, Any]:
    """Convert one :class:`ParameterSchema` node to a standards-compliant JSON Schema object.

    The per-property ``required`` boolean is **never** emitted; the parent object collects
    required names into an array instead. ``items`` is always normalised to a schema object.
    """
    base: Dict[str, Any] = {"type": schema.type}

    if schema.description is not None:
        base["description"] = schema.description
    if schema.enum is not None:
        base["enum"] = list(schema.enum)

    if schema.type == "array":
        raw = schema.items
        item_schema = ParameterSchema(raw) if isinstance(raw, str) else raw
        # ``items`` is guaranteed non-None for arrays by the registry validator.
        base["items"] = to_json_schema(item_schema) if item_schema is not None else {}
    elif schema.type == "object":
        if schema.properties is not None:
            properties: Dict[str, Any] = {}
            req: List[str] = []
            for key, prop in schema.properties.items():
                properties[key] = to_json_schema(prop)
                if prop.required:
                    req.append(key)
            base["properties"] = properties
            if req:
                base["required"] = req

    return base


def params_to_json_schema(params: Dict[str, ParameterSchema]) -> Dict[str, Any]:
    """Build a top-level object schema from a flat map of named :class:`ParameterSchema`."""
    properties: Dict[str, Any] = {}
    req: List[str] = []
    for name, schema in params.items():
        properties[name] = to_json_schema(schema)
        if schema.required:
            req.append(name)

    result: Dict[str, Any] = {"type": "object", "properties": properties}
    if req:
        result["required"] = req
    return result


__all__ = ["to_json_schema", "params_to_json_schema"]

File: bindings/dotnet/Kriya/JsonSchema.cs
using System.Text.Json.Nodes;

namespace Kriya;

/// <summary>
/// Standards-compliant JSON Schema (draft 2020-12 clean) export. kriya's compact
/// <see cref="ParameterSchema"/> uses a per-property <c>Required</c> bool and may carry a bare item
/// type; MCP clients want <c>required</c> as an array on the parent object and <c>items</c> as a
/// schema object. This converts between them, byte-faithful to kriya-core's <c>jsonschema.ts</c>.
/// </summary>
public static class JsonSchema
{
    internal static string TypeName(ParamType t) => t switch
    {
        ParamType.String => "string",
        ParamType.Number => "number",
        ParamType.Boolean => "boolean",
        ParamType.Array => "array",
        ParamType.Object => "object",
        _ => "object",
    };

    /// <summary>Convert one <see cref="ParameterSchema"/> node to a JSON Schema object.</summary>
    public static JsonObject ToJsonSchema(ParameterSchema schema)
    {
        var node = new JsonObject { ["type"] = TypeName(schema.Type) };

        if (schema.Description is not null)
            node["description"] = schema.Description;

        if (schema.Enum is not null)
        {
            var arr = new JsonArray();
            foreach (var v in schema.Enum)
                arr.Add(JsonValue.Create(v));
            node["enum"] = arr;
        }

        switch (schema.Type)
        {
            case ParamType.Array:
                // Items is guaranteed non-null for arrays by the registry validator.
                node["items"] = schema.Items is not null ? ToJsonSchema(schema.Items) : new JsonObject();
                break;

            case ParamType.Object when schema.Properties is not null:
                var props = new JsonObject();
                var required = new JsonArray();
                foreach (var (key, prop) in schema.Properties)
                {
                    props[key] = ToJsonSchema(prop);
                    if (prop.Required) required.Add((JsonNode)key);
                }
                node["properties"] = props;
                if (required.Count > 0) node["required"] = required;
                break;
        }

        return node;
    }

    /// <summary>Build a top-level object schema from a flat map of named parameter schemas.</summary>
    public static JsonObject ParamsToJsonSchema(IReadOnlyDictionary<string, ParameterSchema> parameters)
    {
        var props = new JsonObject();
        var required = new JsonArray();
        foreach (var (name, schema) in parameters)
        {
            props[name] = ToJsonSchema(schema);
            if (schema.Required) required.Add((JsonNode)name);
        }

        var result = new JsonObject { ["type"] = "object", ["properties"] = props };
        if (required.Count > 0) result["required"] = required;
        return result;
    }
}
